ttnn/cpp/ttnn/operations/moreh/moreh_adamw/device/moreh_adamw_device_operation.cpp

```diff
@@ -0,0 +1,134 @@
+        tt::operations::primary::check_tensor(
+            tensor_args.max_exp_avg_sq_in.value(), "moreh_adamw", "max_exp_avg_sq_in");
+    }
+
+    if (tensor_args.param_out.has_value()) {
+        tt::operations::primary::check_tensor(tensor_args.param_out.value(), "moreh_adamw", "param_out");
+    }
+    if (tensor_args.exp_avg_out.has_value()) {
+        tt::operations::primary::check_tensor(tensor_args.exp_avg_out.value(), "moreh_adamw", "exp_avg_out");
+    }
+    if (tensor_args.exp_avg_sq_out.has_value()) {
+        tt::operations::primary::check_tensor(tensor_args.exp_avg_sq_out.value(), "moreh_adamw", "exp_avg_sq_out");
+    }
+    if (tensor_args.max_exp_avg_sq_out.has_value()) {
+        tt::operations::primary::check_tensor(
+            tensor_args.max_exp_avg_sq_out.value(), "moreh_adamw", "max_exp_avg_sq_out");
+    }
+}
+
+void MorehAdamWDeviceOperation::validate_on_program_cache_hit(
+    const operation_attributes_t& attributes, const tensor_args_t& tensor_args) {}
+
+MorehAdamWDeviceOperation::shape_return_value_t MorehAdamWDeviceOperation::compute_output_shapes(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    auto input_tensor_shape = tensor_args.param_in.get_shape();
+
+    return {
+        input_tensor_shape,
+        input_tensor_shape,
+        input_tensor_shape,
+        input_tensor_shape,
+    };
+}
+
+MorehAdamWDeviceOperation::tensor_return_value_t MorehAdamWDeviceOperation::create_output_tensors(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    auto output_shapes = compute_output_shapes(operation_attributes, tensor_args);
+    auto dtype = tensor_args.param_in.get_dtype();
+    Layout layout{Layout::TILE};
+    auto device = tensor_args.param_in.device();
+
+    tensor_return_value_t result;
+
+    if (tensor_args.param_out.has_value()) {
+        result.push_back(tensor_args.param_out.value());
+    } else {
+        result.push_back(
+            create_device_tensor(output_shapes.at(0), dtype, layout, device, operation_attributes.mem_config));
+    }
+
+    if (tensor_args.exp_avg_out.has_value()) {
+        result.push_back(tensor_args.exp_avg_out.value());
+    } else {
+        result.push_back(
+            create_device_tensor(output_shapes.at(1), dtype, layout, device, operation_attributes.mem_config));
+    }
+
+    if (tensor_args.exp_avg_sq_out.has_value()) {
+        result.push_back(tensor_args.exp_avg_sq_out.value());
+    } else {
+        result.push_back(
+            create_device_tensor(output_shapes.at(2), dtype, layout, device, operation_attributes.mem_config));
+    }
+
+    if (tensor_args.max_exp_avg_sq_out.has_value()) {
+        result.push_back(tensor_args.max_exp_avg_sq_out.value());
+    } else {
+        result.push_back(
+            create_device_tensor(output_shapes.at(3), dtype, layout, device, operation_attributes.mem_config));
+    }
```

Comment:
```suggestion
    if (tensor_args.max_exp_avg_sq_out.has_value()) {
        result.push_back(tensor_args.max_exp_avg_sq_out.value());
    } else if(operation_attributes.amsgrad) {
        result.push_back(
            create_device_tensor(output_shapes.at(3), dtype, layout, device, operation_attributes.mem_config));
    }
```
`max_exp_avg_sq_out` is only needed when `amsgrad` is True, so it should only be created in that case

Reply:
Updated